src/main/java/com/microsoft/sqlserver/jdbc/SQLServerException.java

```diff
@@ -360,8 +360,19 @@ static String generateStateCode(SQLServerConnection con, int errNum, int databas
                     return "40001"; // deadlock detected
                 case 2627:
                     return "23000"; // DPM 4.04. Primary key violation
-                default:
-                    return "S000" + databaseState;
+                default: {
+                    String dbState = databaseState.toString();
+                    /*
+                     * Length allowed for SQL State is 5 characters as per SQLSTATE specifications. Append trailing
+                     * zeroes as needed based on length of database error State as length of databaseState is either 1
+                     * or 2 characters.
+                     */
+                    StringBuilder trailingZeroes = new StringBuilder("S");
+                    for (int i = 0; i < 4 - dbState.length(); i++) {
+                        trailingZeroes.append("0");
+                    }
```

Comment:
`String.format("S%4s", s).replace(' ', '0')` pads 0's to the left of the string up to a length of 4 and appends uppercase 'S' to the beginning. The default block can just become:
```java
default:
    return String.format("S%4s", databaseState.toString()).replace(' ', '0');
```

Reply:
String.format is not good at performance when compared with StringBuilder, hence not implemented.

For negative values test scenario, the negative values if passed to SQL Server with RAISEERROR, are received from SQL Server as positive integer 1, hence no check is needed.